clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -0,0 +1,359 @@
+          .stream()
+          .collect(Collectors.toMap(e -> Integer.valueOf(e.getKey().toString()), e -> e.getValue().toString()));
+      Map<Integer, String> valueSchemas = ((HashMap<Utf8, Utf8>) metadataResponse.get("valueSchemas")).entrySet()
+          .stream()
+          .collect(Collectors.toMap(e -> Integer.valueOf(e.getKey().toString()), e -> e.getValue().toString()));
+      // Map<Utf8, Integer> helixGroupInfo = (HashMap<Utf8, Integer>) metadataResponse.get("helixGroupInfo");
+      Map<Integer, List<String>> routingInfo =
+          ((HashMap<Utf8, Collection<Utf8>>) metadataResponse.get("routingInfo")).entrySet()
+              .stream()
+              .collect(
+                  Collectors.toMap(
+                      e -> Integer.valueOf(e.getKey().toString()),
+                      e -> e.getValue().stream().map(Utf8::toString).collect(Collectors.toList())));
+
+      if (fetchedVersion != getCurrentStoreVersion()) {
+        // call the DICTIONARY endpoint if needed
+        CompletableFuture<TransportClientResponse> dictionaryFetchFuture = null;
+        if (!versionZstdDictionaryMap.containsKey(fetchedVersion)
+            && compressionStrategy == CompressionStrategy.ZSTD_WITH_DICT.getValue()) {
+          dictionaryFetchFuture = fetchCompressionDictionary(fetchedVersion);
+        }
+
+        // Update partitioner pair map (versionPartitionerMap)
+        Properties params = new Properties();
+        params.putAll(partitionerParams);
+        VenicePartitioner partitioner =
+            PartitionUtils.getVenicePartitioner(partitionerClass, amplificationFactor, new VeniceProperties(params));
+        versionPartitionerMap.put(fetchedVersion, new PartitionerPair(partitioner, partitionCount));
+
+        // Update readyToServeInstanceMap
+        for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+          String key = getVersionPartitionMapKey(fetchedVersion, partitionId);
+          readyToServeInstancesMap.put(key, routingInfo.get(partitionId));
+        }
+
+        // Update schemas
+        SchemaData schemaData = new SchemaData(storeName);
+        for (Map.Entry<Integer, String> entry: keySchema.entrySet()) {
+          schemaData.setKeySchema(new SchemaEntry(entry.getKey(), entry.getValue()));
+        }
+        for (Map.Entry<Integer, String> entry: valueSchemas.entrySet()) {
+          schemaData.addValueSchema(new SchemaEntry(entry.getKey(), entry.getValue()));
+        }
+        schemas.set(schemaData);
+
+        // Evict entries that are over two versions behind from the latest fetched version
+        readyToServeInstancesMap.entrySet().removeIf(entry -> getVersionFromKey(entry.getKey()) < fetchedVersion - 2);
+        versionPartitionerMap.entrySet().removeIf(entry -> entry.getKey() < fetchedVersion - 2);
+        versionZstdDictionaryMap.entrySet().removeIf(entry -> entry.getKey() < fetchedVersion - 2);
```

Comment:
Hardcoding to 2 versions is incorrect since failed pushes will lead to invalid states. We should see which versions are active in Venice and remove the inactive ones

Reply:
I'm not sure if I have the current active versions so I think I might need to modify my endpoint to send this information along. Would this just be the result from `Store.getVersions()`?

Reply:
Made the change to only evict versions that aren't in our active versions set